RavenDB-14223 time-series rollups & retention

karmeli87 opened the pull request.

karmeli87 committed:
RavenDB-14223 implement the initial API

karmeli87 committed:
RavenDB-14223 wire server side for editing the configuration

karmeli87 committed:
RavenDB-14223 wire the client to configure the time-series

ayende commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesCollectionConfiguration.cs:
```diff
@@ -0,0 +1,181 @@
+        /// <summary>
+        /// How far from UTC.Now this policy should be applied
+        /// </summary>
+        public TimeSpan TimeFromNow;
```
I don't understand this.
You have two sides here: What the time frame that we work on? And what is the frequency that we do the roll up.
But the API seems really confusing. Also, why not simple have two time spans here?

karmeli87 replied:
This draft PR is exactly for this purpose, to decide on a good API.
 
I need here some ASCII art to explain what I've tried to do:
`+---- raw -----|t1|----- rollup1 ------|t2|------ rollup2 -----| delete >`

Each rollup policy have `t`, which defines how far from `now` is should be applied.
Also, it defines the "aggregation factor" which define the down sampling..

ayende commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesCollectionConfiguration.cs:
```diff
@@ -0,0 +1,181 @@
+
+        public bool Disabled;
+
+        public List<TimeSeriesDownSamplePolicy> DownSamplePolicies;
```
Better to call it roll up, not down sampling

ayende commented on src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:
```diff
@@ -741,6 +741,24 @@ public void WriteDatabaseRecord(DatabaseRecord databaseRecord, SmugglerProgressB
+                    {
+                        foreach (var collection in currentDatabaseRecord.TimeSeries.Collections)
+                        {
+                            if ((databaseRecord.TimeSeries.Collections.ContainsKey(collection.Key)) == false)
```
Why this check?

karmeli87 replied:
This is that check we do for revision configuration as well. 
I guess we don't want to override an existing configuration? But I agree that it might result in an unexpected behavior without any warning/message. 
What do we want to do here? overwrite or notify that we skip this config, because one is already exists?

ayende replied:
I think this was originally meant for import, so if you update an existing value?

ayende commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesCollectionConfiguration.cs:
```diff
@@ -0,0 +1,181 @@
+            if (DownSamplePolicies.Count == 0)
+                return;
+
+            DownSamplePolicies.Sort(TimeSeriesDownSamplePolicyComparer.Instance);
```
This needs to run server side as well

karmeli87 committed:
RavenDB-14223 better API

karmeli87 committed:
RavenDB-14223 implement a single node - good path roll up

karmeli87 committed:
RavenDB-14223 handle distributed rollups

karmeli87 committed:
RavenDB-14223 better client API

karmeli87 committed:
RavenDB-14223 move time-series stats to a table

This change will allow an easy query on the the time-series by policy.
So we can pick the relevant time-series upon roll up configuration change.

karmeli87 committed:
RavenDB-14223 handle adding/removing policies

karmeli87 committed:
RavenDB-14223 place time-series slice holder in a dedicated class

karmeli87 committed:
RavenDB-14223 move start/end to be part of the stats table

karmeli87 committed:
RavenDB-14223 implement time-series retention

ayende commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesCollectionConfiguration.cs:
```diff
@@ -40,12 +41,18 @@ internal RollupPolicy GetPolicyByTimeSeries(string name)
 
         internal RollupPolicy GetNextPolicy(RollupPolicy policy)
```
This looks rather awkward API.
Why not simply do enumeration?

ayende commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesCollectionConfiguration.cs:
```diff
@@ -0,0 +1,202 @@
+            return RollupPolicies[current + 1];
+        }
+
+        internal RollupPolicy GetPreviousPolicy(RollupPolicy policy)
```
I don't like this kind of API

karmeli87 committed:
RavenDB-14223 raw policy can't be static

and change the policy class name

ayende commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs:
```diff
@@ -0,0 +1,655 @@
+                return runner;
+            }
+            catch (Exception e)
+            {
```
Make sure that you dispose the old `policyRunner` here

ayende commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs:
```diff
@@ -0,0 +1,655 @@
+
+        public unsafe void MarkForPolicy(DocumentsOperationContext context, Slice key, string collection, string name, long etag, DateTime baseline, string changeVector)
+        {
+            var config = Configuration.Collections[collection];
```
What if this does not exists?

ayende commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesCollectionConfiguration.cs:
```diff
@@ -0,0 +1,202 @@
+            return RollupPolicies.SingleOrDefault(p => string.Compare(p.Name,policy, StringComparison.InvariantCultureIgnoreCase) == 0);
+        }
+
+        internal RollupPolicy GetPolicyByTimeSeries(string name)
```
This is very strange way to find a policy by name.
Why are you searching for a greater than here in an unordered policies?

karmeli87 committed:
RavenDB-14223 make retention and rollup work together

karmeli87 committed:
RavenDB-14223 add parameters validation

ppekrol commented on src/Raven.Client/Documents/Operations/TimeSeries/ConfigureTimeSeriesOperation.cs:
```diff
@@ -0,0 +1,70 @@
+
+        public ConfigureTimeSeriesOperation(TimeSeriesConfiguration configuration)
+        {
+            _configuration = configuration;
```
add `null` check

karmeli87 replied:
`null` should remove the configuration and close the runner. 
will verify and add a test

ppekrol commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesCollectionConfiguration.cs:
```diff
@@ -0,0 +1,229 @@
+    public class TimeSeriesCollectionConfiguration : IDynamicJson
+    {
+
+        public bool Disabled;
```
Let's use properties here. We do not care about performance here and you can easily override properties in the future, and fields are not that easy to work with.

ppekrol commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesCollectionConfiguration.cs:
```diff
@@ -0,0 +1,229 @@
+            return new DynamicJsonValue
+            {
+                [nameof(Policies)] = new DynamicJsonArray(Policies.Select(p=>p.ToJson())),
+                [nameof(RawPolicy)] = RawPolicy.ToJson(),
```
Can this be null?

karmeli87 replied:
no, test added

ppekrol commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesCollectionConfiguration.cs:
```diff
@@ -0,0 +1,229 @@
+        {
+            return new DynamicJsonValue
+            {
+                [nameof(Policies)] = new DynamicJsonArray(Policies.Select(p=>p.ToJson())),
```
Policies can be null?

karmeli87 replied:
no, test added

ppekrol commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesCollectionConfiguration.cs:
```diff
@@ -0,0 +1,229 @@
+    public class RawTimeSeriesPolicy : TimeSeriesPolicy
+    {
+        internal const string PolicyString = "rawpolicy"; // must be lower case
+        public RawTimeSeriesPolicy()
```
IIRC you can make it internal?

karmeli87 replied:
Now I actually using it.

ppekrol commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesCollectionConfiguration.cs:
```diff
@@ -0,0 +1,229 @@
+            // for de-serializer
+        }
+
+        public RawTimeSeriesPolicy(TimeSpan? retentionTime)
```
Sanity-check. Null is expected value here?

karmeli87 replied:
removed the `nullable` and the ambiguity between `null` and `TimeSpan.MaxValue`.

ppekrol commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesCollectionConfiguration.cs:
```diff
@@ -0,0 +1,229 @@
+        /// Define the aggregation type
+        /// </summary>
+        public AggregationType Type { get; private set; }
+        // TODO: consider Continuous Query approach
```
TODO

ppekrol commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesCollectionConfiguration.cs:
```diff
@@ -0,0 +1,229 @@
+            Type = type;
+
+            var retentionStr = retentionTime == TimeSpan.MaxValue ? "" : $"KeepFor{retentionTime}";
+            Name = $"Every{AggregationTime}By{Type}{retentionStr}";
```
Do we want to 'split' it using `_` like in index names?

ppekrol commented on src/Raven.Client/Documents/Queries/TimeSeries/TimeSeriesAggregation.cs:
```diff
@@ -25,4 +25,15 @@ public class TimeSeriesRangeAggregation
+        Min,
+        Max,
+        Mean,
+        Avg,
```
rename to `Average` it will match what we have in Facets

ayende commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesCollectionConfiguration.cs:
```diff
@@ -0,0 +1,229 @@
+            if (policy == RawTimeSeriesPolicy.PolicyString)
+                return RawPolicy;
+
+            return Policies.SingleOrDefault(p => string.Compare(p.Name,policy, StringComparison.InvariantCultureIgnoreCase) == 0);
```
I don't understand this. Why are you doing a query like that, and why is it `O(n)` ?

karmeli87 replied:
The policies sorted by `AggregationTime`, I can extract this time from the name. 
But should we care about it? `O(n)` where `n` is expected to be low.
What do you suggest? I really don't want to persist the index of the policy.

ayende commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesCollectionConfiguration.cs:
```diff
@@ -0,0 +1,229 @@
+            if (name.Contains(TimeSeriesConfiguration.TimeSeriesRollupSeparator) == false)
+                return RawPolicy;
+
+            return Policies.SingleOrDefault(p => name.IndexOf(p.Name, StringComparison.InvariantCultureIgnoreCase) > 0);
```
Why comparing by greater than?

karmeli87 replied:
If it contains or not? and `Contains` doesn't work with `StringComparison`.

ayende commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesCollectionConfiguration.cs:
```diff
@@ -0,0 +1,229 @@
+            if (policy == RawPolicy)
+                return Policies[0];
+
+            var current = Policies.FindIndex(p => p == policy);
```
This is pretty ugly. Can't we hold the index of the policy, so we don't have to do this?

karmeli87 replied:
Okay, let's hold the index

ayende commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesCollectionConfiguration.cs:
```diff
@@ -0,0 +1,229 @@
+            if (policy == RawPolicy)
+                return TimeSeriesPolicy.BeforeAllPolices;
+
+            var current = Policies.FindIndex(p => p == policy);
```
Same, enumerator, maybe?

ayende commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesCollectionConfiguration.cs:
```diff
@@ -0,0 +1,229 @@
+        /// <summary>
+        /// Define the aggregation type
+        /// </summary>
+        public AggregationType Type { get; private set; }
```
If we do this, then we have to provide the ability to have multiple rollups. One for min, one for max, etc.
Can't we do all the aggregations, instead?

karmeli87 replied:
This actually an interesting issue.
Let's say we have a single value time-series. Stock price. 

Right now each aggregation result is represented by a point on the policy time-series. So if I aggregate daily by `Average`, each day I'll create a single point from the raw stock price.

But if I want to aggregate daily by `First`, `Last`, `High`, `Low`.
Right now it is a problem, since I can define only a single type.
To allow this kind of thing we can either create a complex policy tree or create a single 4-dim time-series (which I like more).

But do we want to tackle this now?

karmeli87 replied:
The raw time series will be always aggregated into (`First`,`Last`,`Min`,`Max`,`Sum`,`Count`).
In that order.
Other policies will aggregate each of those values further independently.

ppekrol commented on src/Raven.Server/Documents/TimeSeries/BitsBuffer.cs:
```diff
@@ -123,6 +123,7 @@ public BitsBuffer(byte* buffer, int size)
             Header = (BitsBufferHeader*)buffer;
             _buffer = buffer + sizeof(BitsBufferHeader) + Header->CompressedSize;
             Size = size;
+            Debug.Assert(size > 0);
```
Write a message here. It can be as simple as `size > 0`. It is easier to trace failing assert then

ppekrol commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs:
```diff
@@ -0,0 +1,683 @@
+            // documentId/Name
+            Key,
+            Collection,
+            NextRollUp,
```
Isn't the word just `Rollup` ?

ppekrol commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs:
```diff
@@ -0,0 +1,683 @@
+            using (StorageEnvironment.GetStaticContext(out var ctx))
+            {
+                Slice.From(ctx, "TimeSeriesRollUps", ByteStringType.Immutable, out TimeSeriesRollUps);
+                Slice.From(ctx, "RollUpKey", ByteStringType.Immutable, out RollUpKey);
```
nameof?

ppekrol commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs:
```diff
@@ -0,0 +1,683 @@
+
+        public unsafe void MarkForPolicy(DocumentsOperationContext context, Slice key, string collection, string name, long etag, DateTime baseline, string changeVector)
+        {
+            var config = Configuration.Collections[collection];
```
Are we certain that this MUST be in that dictionary? Shouldn't we use TryGet?

ppekrol commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs:
```diff
@@ -0,0 +1,683 @@
+            var table = context.Transaction.InnerTransaction.OpenTable(RollUpSchema, TimeSeriesRollUps);
+            using (table.Allocate(out var tvb))
+            using (Slice.From(context.Allocator, key.Content.Ptr, key.Size - sizeof(long) - sizeof(byte), ByteStringType.Immutable, out var timeSeriesKey))
+            using (Slice.From(context.Allocator,collection, ByteStringType.Immutable,out var collectionSlice))
```
This should be lowered?

ayende replied:
Yes, collection names are case insensitive

karmeli87 replied:
So we have a bigger problem here, we keep it with the original casing
e.g. https://github.com/ravendb/ravendb/blob/23c596b3b0f84bca9f7c19ec02e64a89850189a2/src/Raven.Server/Documents/TimeSeries/TimeSeriesStorage.cs#L906
Also, why do we store the collection at all? for replication only? 

Anyway, why do we care about it here? we need only to ensure that when we read from the configuration we do it in case-insensitive matter.

ppekrol commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs:
```diff
@@ -0,0 +1,683 @@
+            using (table.Allocate(out var tvb))
+            using (Slice.From(context.Allocator, key.Content.Ptr, key.Size - sizeof(long) - sizeof(byte), ByteStringType.Immutable, out var timeSeriesKey))
+            using (Slice.From(context.Allocator,collection, ByteStringType.Immutable,out var collectionSlice))
+            using (Slice.From(context.Allocator, nextPolicy.Name, ByteStringType.Immutable, out var policyToApply))
```
This should be lowered?

ppekrol commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs:
```diff
@@ -0,0 +1,683 @@
+        {
+            var topology = _database.ServerStore.LoadDatabaseTopology(_database.Name);
+            var isFirstInTopology = string.Equals(topology.Members.FirstOrDefault(), _database.ServerStore.NodeTag, StringComparison.OrdinalIgnoreCase);
+            if (isFirstInTopology == false)
```
Do we want to go with first? Isn't that the most busiest one? It probably it the most up to date, but do we care about that?

ayende replied:
This is what we do for expiration and refreshing. That give us the most consistency

karmeli87 replied:
But there is a point that we are putting more and more load on the primary.. for now I leave it as is.

ppekrol commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs:
```diff
@@ -0,0 +1,683 @@
+            }
+        }
+
+        public static void SplitKey(Slice key, out string docId, out string name)
```
Sanity-check. Shouldn't be an issue that we are allocating here instead of 'slicing' Slices?

ayende replied:
I think that this uses implicit conversions. But should it?

karmeli87 replied:
We are indeed allocating here. 
But let's optimize it a bit later, since changing this will require to change `string` into `slice` in various places

ppekrol commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs:
```diff
@@ -0,0 +1,683 @@
+
+        internal class TimeSeriesRetentionCommand : TransactionOperationsMerger.MergedTransactionCommand
+        {
+            public const int BatchSize = 1024;
```
What is this used for?

ayende replied:
Used elsewhere to limit how many items we scan per batch

ppekrol commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs:
```diff
@@ -0,0 +1,683 @@
+                    using (Slice.From(context.Allocator, _to.Name, ByteStringType.Immutable, out var policyToApply))
+                    using (Slice.From(context.Allocator, string.Empty, ByteStringType.Immutable, out var changeVectorSlice))
+                    {
+                        tvb.Add(key);
```
Are we sure we want code like this here instead of e.g. `TimeSeriesStorage.Rollups.MethodName()` ?

karmeli87 replied:
Not sure I'm following.

ppekrol replied:
What I've meant is that we are using xxxStorage classes for storage interaction and here you are inserting directly into table without using TimeSeriesStorage.Rollups for example

karmeli87 replied:
Okay, I will move it to dedicated class.

ppekrol commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs:
```diff
@@ -0,0 +1,683 @@
+
+            public override TransactionOperationsMerger.IReplayableCommandDto<TransactionOperationsMerger.MergedTransactionCommand> ToDto(JsonOperationContext context)
+            {
+                throw new NotImplementedException();
```
Remember that we need to fix this later on

ppekrol commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesStats.cs:
```diff
@@ -0,0 +1,290 @@
+        {
+            using (StorageEnvironment.GetStaticContext(out ByteStringContext ctx))
+            {
+                Slice.From(ctx, "TimeSeriesStats", ByteStringType.Immutable, out TimeSeriesStatsKey);
```
nameof?

ppekrol commented on src/Raven.Server/ServerWide/Commands/EditTimeSeriesConfigurationCommand.cs:
```diff
@@ -0,0 +1,39 @@
+            {
+                foreach (var config in Configuration.Collections.Values)
+                {
+                    config.Validate();
```
can value be null?

karmeli87 replied:
maybe. will safe-guard.

ayende commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesCollectionConfiguration.cs:
```diff
@@ -0,0 +1,229 @@
+            Type = type;
+
+            var retentionStr = retentionTime == TimeSpan.MaxValue ? "" : $"KeepFor{retentionTime}";
+            Name = $"Every{AggregationTime}By{Type}{retentionStr}";
```
This will generate names such as: "Every00:30ByAvgKeepFor1.00.00", no?
That seems pretty ugly

karmeli87 replied:
Yes. Sorry. 
Should we let the user define it? 

Note that I persist the name of the policy which is unique for it's definition.

ppekrol replied:
Can we convert those timespan values to some human readable value and use them in the name or do we prefer user to specify the name?

karmeli87 replied:
We will let the user to define the policy name. It should be unique and we will do only best effort validation.

ayende commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs:
```diff
@@ -0,0 +1,683 @@
+                    if (Equals(policyRunner.Configuration, dbRecord.TimeSeries))
+                        return policyRunner;
+                }
+                policyRunner?.Dispose();
+                var runner = new TimeSeriesPolicyRunner(database, dbRecord.TimeSeries);
+                runner.Start();
```
Do we have to worry about concurrency here?

karmeli87 replied:
Don't believe so, since eventually all goes through the tx merger.

ayende commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs:
```diff
@@ -0,0 +1,683 @@
+        public static readonly Slice TimeSeriesRollUps;
+        private static readonly Slice RollUpKey;
+        private static readonly Slice NextRollUpIndex;
+        private enum RollUpColumn
```
```suggestion
        private enum RollUpColumns
```

ayende commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs:
```diff
@@ -0,0 +1,683 @@
+
+            var current = config.GetPolicyByTimeSeries(name);
+            if (current == null) // policy not found
+                // TODO: delete this time-series if not the raw one?
```
No, changing the policy doesn't delete data

karmeli87 replied:
So changing the configuration should keep left-overs?
The user should manually delete them?

karmeli87 replied:
Removing the policy will keep the data by default and it will be user's responsibility to delete them.
 
Changing the parameters for an existing policy will take effect from the time of the change and will keep previous data intact. (unless is should re-aggregate due to the parameters change)

ayende commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs:
```diff
@@ -0,0 +1,683 @@
+                // TODO: delete this time-series if not the raw one?
+                return;
+
+            var nextPolicy = config.GetNextPolicy(current);
```
I really don't like the next / prev thing, what purpose does it serve?

ayende commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs:
```diff
@@ -0,0 +1,683 @@
+                        if (command.RolledUp == 0)
+                            break;
+
+                        states.Clear();
```
Shouldn't this be at the top ? Just in case we change the structure, it would be easier to manage

ayende commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs:
```diff
@@ -0,0 +1,683 @@
+                        {
+                            Collection = collection,
+                            From = DateTime.MinValue
+                        };
```
Shouldn't you have a `To` here?

karmeli87 replied:
I populate this in line `394` .
Each policy has a different to.

ayende commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs:
```diff
@@ -0,0 +1,683 @@
+                    _request.Name = name;
+                    _request.DocumentId = docId;
+
+                    context.DocumentDatabase.DocumentsStorage.TimeSeriesStorage.RemoveTimestampRange(context, _request);
```
I think this will delete everything, because `To` isn't specified.

ayende commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs:
```diff
@@ -0,0 +1,683 @@
+                RollUpSchema.Create(context.Transaction.InnerTransaction, TimeSeriesRollUps, 16);
+                var table = context.Transaction.InnerTransaction.OpenTable(RollUpSchema, TimeSeriesRollUps);
+
+                var toDelete = new List<Slice>(tss.Stats.GetTimeSeriesNameByPolicy(context, _collection, _policy, 0, BatchSize));
```
What if there are more than `BatchSize`?

karmeli87 replied:
Here we are in the tx merger. We will release it and loop in again from the caller.

ayende commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs:
```diff
@@ -0,0 +1,683 @@
+                    if (_configuration == null)
+                        return RolledUp;
+
+                    var config = _configuration.Collections[item.Collection];
```
TryGet, no?

ayende commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs:
```diff
@@ -0,0 +1,683 @@
+                    var intoTimeSeries = policy.GetTimeSeriesName(rawTimeSeries);
+                    
+                    var intoReader = tss.GetReader(context, item.DocId, intoTimeSeries, startFrom, startFrom);
+                    var previouslyAggregated = intoReader.AllValues().Any();
```
Doesn't this consume the reader?

karmeli87 replied:
Yes, but I need it only for this purpose.

ayende commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs:
```diff
@@ -0,0 +1,683 @@
+                        }
+                    }
+
+                    if (_isFirstInTopology == false)
```
Why do we have to get to this point to bail out? To remove the policies markers?

karmeli87 replied:
Yes, remove rollups that were already done and replicated.

ayende commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesStats.cs:
```diff
@@ -0,0 +1,290 @@
+        {
+            Key = 0, // documentId, separator, name
+            PolicyName = 1,
+            Separator = 2,
```
Why is this its own dedicated column?

karmeli87 replied:
Easier to work with, instead of adding/removing the separator each time.
Is it a problem?

ayende commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesStats.cs:
```diff
@@ -0,0 +1,290 @@
+        private static readonly Slice StartTimeIndex;
+        private static readonly TableSchema TimeSeriesStatsSchema = new TableSchema();
+
+        private enum StatsColumn
```
```suggestion
        private enum StatsColumns
```

ayende commented on src/Sparrow.Server/ByteString.cs:
```diff
@@ -332,6 +332,17 @@ public string Substring(int length)
+            EnsureIsNotBadPointer();
+
+            var encoding = Encodings.Utf8;
+            return encoding.GetString(_pointer->Ptr + offset, length);
```
This is not a valid way to handle things. `offset` might fall inside of UTF8 mulit byte char

karmeli87 replied:
Okay, I'll remove this entirely, since we want to optimize the caller function anyway (`SplitKey`)

karmeli87 committed:
RavenDB-14223 address PR comments

karmeli87 committed:
RavenDB-14223 use average instead of avg

But keep the support for avg() in the RQL

karmeli87 committed:
RavenDB-14223 use name to index cache

ppekrol commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs:
```diff
@@ -0,0 +1,706 @@
+            if (isFirstInTopology == false)
+                return;
+
+            var now = DateTime.UtcNow;
```
database.Time.UtcNow()

karmeli87 committed:
RavenDB-14223 move rollups to dedicated class

karmeli87 committed:
RavenDB-14223 rollup in equidistant intervals

karmeli87 committed:
RavenDB-14223 use the word Rollup

karmeli87 committed:
RavenDB-14223 implement recording for the new commands

karmeli87 committed:
RavenDB-14223 address rollup PR comments

- can specify rollup name
- mark for rollup based on the values instead of the segment
- raw data will be aggregated into (first, last, min, max, sum, count)

karmeli87 committed:
RavenDB-14223 address retention PR comments

karmeli87 committed:
RavenDB-14223 adjustments

ppekrol commented on src/Raven.Client/Documents/Operations/TimeSeries/ConfigureTimeSeriesOperation.cs:
```diff
@@ -0,0 +1,70 @@
+
+        public ConfigureTimeSeriesOperation(TimeSeriesConfiguration configuration)
+        {
+            _configuration = configuration ?? new TimeSeriesConfiguration();
```
Haven't you wrote that setting null here will remove the configuration? `ConfigureExpirationOperation` allows that.

karmeli87 replied:
Actually passing `null` to `ConfigureExpirationOperation` will throw 
```
System.InvalidOperationException : Cannot perform ReadValue when encountered the ReadObjectDocument state
   at Sparrow.Json.ManualBlittableJsonDocumentBuilder`1.ThrowIllegalStateException(ContinuationState state, String realOperation) in C:\Work\RavenDB-5.0\src\Sparrow\Json\ManualBlittableJsonDocumentBuilder.cs:line 582
   at Sparrow.Json.ManualBlittableJsonDocumentBuilder`1.FinishWritingScalarValue(BuildingState currentState) in C:\Work\RavenDB-5.0\src\Sparrow\Json\ManualBlittableJsonDocumentBuilder.cs:line 566
   at Sparrow.Json.ManualBlittableJsonDocumentBuilder`1.WriteValueNull() in C:\Work\RavenDB-5.0\src\Sparrow\Json\ManualBlittableJsonDocumentBuilder.cs:line 284
   at Raven.Client.Json.BlittableJsonWriter.WriteNull() in C:\Work\RavenDB-5.0\src\Raven.Client\Json\BlittableJsonWriter.cs:line 338
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeValue(JsonWriter writer, Object value, JsonContract valueContract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerProperty)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.Serialize(JsonWriter jsonWriter, Object value, Type objectType)
   at Newtonsoft.Json.JsonSerializer.SerializeInternal(JsonWriter jsonWriter, Object value, Type objectType)
   at Newtonsoft.Json.JsonSerializer.Serialize(JsonWriter jsonWriter, Object value)
   ....
```

ppekrol commented on src/Raven.Client/Documents/Queries/TimeSeries/TimeSeriesAggregation.cs:
```diff
@@ -21,8 +21,19 @@ public class TimeSeriesAggregationResult : TimeSeriesQueryResult
+    {
+        Min,
+        Max,
+        Mean,
```
What is the difference between `Mean` and `Average` ?

ppekrol replied:
Arithmetic Mean IIRC is average, is this another type of mean?

karmeli87 replied:
Indeed Arithmetic Mean is Average. We don't support anything else.

ppekrol replied:
I think that we can remove 'Mean' ?
